### PR TITLE
Feature/circuit drawer

### PR DESCRIPTION
**Pull request by jackturnerer**
I have added a custom bullseye graphic for CNOT and CCNOT, a custom measurement graphic, empty dots for anticontrols, label for SqrtSWAP gate. Have also added the option to save the circuit by specifying a "filename". Finally, the user can now specify a "theme" for setting the colours of the circuit (default is boring black and white "bw", but "dark" and "qm" are also options).
image

### CODE REVIEW

**Comment by jackturnerer** [comment]
New color themes:

**"qmt" colour scheme**
Coral for unitary gates, dark blue for stave and initalisations, gray for noise, blue for non-unitary gates (from .operators module) and measurement.

image

File: pyquest/drawer.py [hunk 1]
```diff
@@ -1,17 +1,17 @@
-'''
```
Comment (TysonRayJones) [inline pyquest/drawer.py:1]: A lot of these changes seem to be the result of a linter and make inconsequential, stylistic changes. This isn't really nice to do to someone's existing code; it muddies the blame, and CI can auto-lint if it is crucial to a project. PyQuEST does not even have a style guide, so the many of these changes (e.g. changing single-character `'x'` to `"x"` - hell, that's not even in [PEP 8](https://peps.python.org/pep-0008/#string-quotes)) are completely subjective and a bit pointless

File: pyquest/drawer.py [hunk 2]
```diff
@@ -143,15 +267,13 @@ def get_operated_qubits(gate, num_qubits):
 def get_num_qubits(gates):
 
     # find the biggest indexed target/control qubit among explicitly-targeted gates
-    return 1 + max(
```
Comment (TysonRayJones) [inline pyquest/drawer.py:146]: Some stylistic changes are inconsistent. Here you opt to compact the generator from three lines to one...

File: pyquest/drawer.py [hunk 3]
```diff
@@ -172,21 +294,23 @@ def get_circuit_columns(gates):
...
 
         # all qubits controlled or targeted by gate (global operators return all)
         gate_qubits = get_operated_qubits(gate, num_qubits)
 
         # there must be room for all qubits between those explicitly targeted, for connectors
-        gate_range = (min(gate_qubits), max(gate_qubits)+1)
+        gate_range = (min(gate_qubits), max(gate_qubits) + 1)
 
         # find the leftmost column which fits the gate
-        gate_column = get_gate_column(gate_qubits, columns_of_last_target, columns_occluded_by_connectors)
```
Comment (TysonRayJones) [inline pyquest/drawer.py:189]: while here, you expand the function call from one lines to three! And further...

File: pyquest/drawer.py [hunk 4]
```diff
@@ -445,58 +566,91 @@ def draw_gate(gate, column, num_qubits, plt, ax):
...
+    colors = Colors(theme)
+
+    # Set the background color
+    mpl_figure.patch.set_facecolor(colors.get_fig_color())
+    ax.set_facecolor(colors.get_fig_color())
+
     # draw horizontal qubit stave
     for q in range(num_qubits):
         plt.plot(
-            [-.5, num_columns+.5], [q+.5, q+.5],
```
Comment (TysonRayJones) [inline pyquest/drawer.py:482]: you here arbitrarily split these arguments over two lines, throwing away their semantic connection (that they are lists of corresponding `x` and `y` values, and are ergo distinct from the logically uncoupled kwargs that follow)

File: pyquest/drawer.py [hunk 5]
```diff
@@ -279,163 +431,132 @@ def get_gate_rect_style(gate):
...
     indAndNums = enumerate(sorted(nums))
-    for _, group in groupby(indAndNums, lambda x:x[0]-x[1]):
+    for _, group in groupby(indAndNums, lambda x: x[0] - x[1]):
         yield list(map(itemgetter(1), group))
 
 
 def get_gate_graphic_components(gate, column, num_qubits):
 
     # graphics consist of vertical connector lines, control circles, and gate body rectangles
-    lines = []        # item = [(x0,y0), (x1,y1)]
```
Comment (TysonRayJones) [inline pyquest/drawer.py:365]: (unrelated to the other consecutive comments)

Here you have reduced readability by forcing a convention that a robot might want of its code; to have constant spacing everywhere despite its harm to human readability. C'mon now!

**Review by TysonRayJones — COMMENTED** [review]
Is there a strong case to be made for the stylistic changes?

Edit: Richard and I have discussed the possibility of a pyQuEST style guide, or a CI linter, and how best to handle linting being mixed in with commits. So that we don't have to rush those decisions, I'll accept this PR and the blame muddying. For future commits, I'd suggest we only commit new code and keep linting/stylistic revisions of existing code in separate commits (or better, defer it until the above decisions are made). It's easy enough to commit only [_some_](https://stackoverflow.com/questions/1085162/commit-only-part-of-a-files-changes-in-git) local changes (when IDEs are automatically making changes due to the flake8 config), and super easy if you use the (controversially, imo lovely) [Github Desktop](https://github.com/apps/desktop) application.

**Comment by TysonRayJones** [comment]
As a final TODO - where do we stick the doc for this thing?

**Comment by jackturnerer** [comment]
For docs, could put a little note about the circuit drawer with an example here: https://pyquest.org/userguide.html#quantum-circuits , or under #quantum-operators? I'm happy to write this bit.

**Pull request merged**